### DISCUSSION TITLE
Serverless target specific files not available during plugin execution

### DISCUSSION THREAD

**Discussion by asterikx** [vercel/next.js#16096] [Help]
I'm trying to write a plugin (based on [this code](https://github.com/serverless-nextjs/serverless-next.js/tree/master/packages/libs/lambda-at-edge) from [serverless-next.js](https://github.com/serverless-nextjs/serverless-next.js)) to generate (serverless) handler code for AWS Lambda@Edge.
This plugin needs to read from `serverless` target-specific files, in particular from the `prerender-manifest.json` manifest file.
However, these files seem not yet available during the webpack execution. Does anyone know how/when these can be accessed?

To give a rough idea, usage should look like:

```js
// next.config.js

const withPlugins = require('next-compose-plugins');
const images = require('next-images');
const lambdaEdge = require('./lambda-edge-plugin');

const plugins = [lambdaEdge, images];

const nextConfig = {
  distDir: process.env.BUILD_DIR || '.next',
  target: 'serverless',
};

module.exports = withPlugins(plugins, nextConfig);
```

For this, I created the following `NextLambdaEdgePlugin` webpack plugin:

```js
class NextLambdaEdgePlugin {
  constructor(nextConfig, options) {
    const { config: { distDir } } = nextConfig;
    this.nextDistDir = distDir;
  }

  apply(compiler) {
    const done = (stats) => {
      const { nextDistDir } = this;

      // return generateHanlderCode(...);

      fse.readdir(nextDistDir, (err, files) => {
        files.forEach((file) => {
          console.log(file);
        });
      });
      return Promise.resolve();
    };

    compiler.hooks.done.tapPromise('NextLambdaEdgePlugin', done);
  }
}
```

This prints:
```
build-manifest.json
cache
react-loadable-manifest.json
routes-manifest.json
serverless
static
```

What I need during plugin execution is:
```
BUILD_ID
build-manifest.json
cache
export-marker.json
prerender-manifest.json
react-loadable-manifest.json
routes-manifest.json
serverless
static
```

The `serverless` target-specific files are present after `next build` but not yet during the execution of the plugin. Is there any chance I can access them from the plugin?

**Comment by timneutkens** [comment]
This is expected, you can't access build artifacts in this way.

**Reply by asterikx** [reply]
@timneutkens thanks for the reply.

> This is expected, you can't access build artifacts in this way.

Does this mean that it is not possible to access build artifacts from plugins at all?

After all, my plugin lists all build artifacts, except the serverless target artifacts. Also, the [`@next/bundle-analyzer`](https://github.com/vercel/next.js/tree/canary/packages/next-bundle-analyzer) (which uses `webpack-bundle-analyzer`) runs on the final bundle.

What is the recommended way to execute post-build actions using the final build output (preferably as plugins)?

Thanks
